lib/UR/Role.pm

```diff
@@ -0,0 +1,732 @@
+
+Roles are defined with the C<role> keyword.  Their definition looks very
+similar to a class definition as described in L<UR::Object::Type::Initializer>.
+In particular, Roles have a C<has> section to define properties, and accept
+many class-meta attributes such as 'id_generator', 'valid_signals', and 'doc'.
+
+Roles may implement operator overloading via the 'use overload' mechanism.
+
+Roles also have unique atributes to declare restrictions on their use.
+
+=over 4
+
+=item requires
+
+A listref of property and method names that must appear in any class composing
+the Role.  Properties and methods defined in other roles or parent classes
+can satisfy a requirement.
+
+=item excludes
+
+A listref of Role names that may not be composed together with this Role.
+This is useful to declare incompatibilities between roles.
+
+=back
+
+=head2 Composing a Role
+
+Compose one or more Roles into a class using the 'roles' attribute in a class
+definition.
+
+  class My::Class {
+      roles => ['My::Role', 'Other::Role'],
+      is => ['Parent::Class'],
+      has => ['prop_a','prop_b'],
+  };
+
+Properties and meta-attributes from the Roles get copied into the composing
+class.  Subroutines defined in the Roles' namespaces are imported into the
+class's namespace.  Operator overloads defined in the Roles are applied to
+the class.
+
+=head3 Property and meta-attribute conflicts
+
+An exception is thrown if multiple Roles are composed together that define
+the same property, even if the composing class defines the same property in
+an attempt to override them.
+
+=head3 Method conflicts
+
+An exception is thrown if multiple Roles are composed together that
+define the same subroutine name unless the composing class's namespace also
+has a subroutine with the same name.  The class is expected to implement
+whatever behavior is required, maybe by calling one role's method or the
+other, both methods, neither, or anything else.
+
+To call a function in a role, the function's fully qualified name, including
+the role's package, must be used.
+
+=head3 Overload conflicts
+
+Like with method conflicts, an exception is thrown if multiple Roles are
+composed together that overload the same operator unless the composing
+class also overloads that same operator.
+
+An exception is also thrown if composed roles define incompatible 'fallback'
+behavior.  If a role does not specify 'fallback', or explicity sets it to
+C<undef>, it is compatible with other values.  A Role that sets its 'fallback'
+value to true or false is only compatible with other roles' values of undef
+or the same true or false value.
+
+=head2 __import__
+
+Each time a Role is composed into a class, its C<__import__()> method is
+called.  C<__import__()> is passed two arguments:
+
+=over 4
+
+=item *
+
+The name of the role
+
+=item *
+
+The class metadata object composing the role.
+
+=back
+
+This happens after the class is completely constructed.
+
+=head2 Deferred Values
```

Comment:
I think I see Deferred Values as solving a problem I would call "required properties"; at least for the documentation's example.  Just like you could require methods be implemented maybe you should be able to require properties to be implemented?

```
use UR::Role;

role ObjectDisplayer {
    requires_properties => ['target_object'],
};

class ShowCars {
    roles => ['ObjectDisplayer'],
    has => [
        target_object => { is => 'Car' },
    ],
};
```

Requiring a property has some sort of implicit requirement about metadata since you could just require a `target_object` method and implementing a `target_object` property would satisfy that so maybe the API would need to be more specific.  That's sort of what `defer` is except it feels too strongly decoupled from the property.